src/Tes.Runner/Storage/ResolutionPolicyHandler.cs

```diff
@@ -56,29 +69,29 @@ public class ResolutionPolicyHandler
 
     private async Task<DownloadInfo> CreateDownloadInfoWithStrategyAsync(FileInput input)
     {
-        var uri = await ApplySasResolutionToUrlAsync(input.SourceUrl, input.SasStrategy);
-
         if (string.IsNullOrEmpty(input.FullFileName))
         {
             throw new ArgumentException("A task input is missing the full filename. Please check the task definition.");
         }
 
-        return new DownloadInfo(input.FullFileName, uri);
+        var uri = await ApplySasResolutionToUrlAsync(input.SourceUrl, input.SasStrategy ?? SasResolutionStrategy.None);
+
+        return new DownloadInfo(ExpandEnvironmentVariables(input.FullFileName), uri);
     }
 
-    private async Task<UploadInfo> CreateUploadInfoWithStrategyAsync(FileOutput output)
+    private static async Task<UploadInfo> CreateUploadInfoWithStrategyAsync(FileOutput output)
     {
-        var uri = await ApplySasResolutionToUrlAsync(output.TargetUrl, output.SasStrategy);
+        var uri = await ApplySasResolutionToUrlAsync(output.TargetUrl, output.SasStrategy ?? SasResolutionStrategy.None);
 
-        if (string.IsNullOrEmpty(output.FullFileName))
-        {
-            throw new ArgumentException("A task output is missing the full filename. Please check the task definition.");
-        }
+        return new UploadInfo(ExpandEnvironmentVariables(output.FullFileName!), uri);
```

Comment:
Since most of the uploads/downloads happen outside of the executor's view of the filesystem, the files that go into/come from the executor need to be prefixed. But, when creating the file, there's no way to know exactly what the prefix is, but there is an environment variable that gets us 99% of the way there.

Reply:
Unless now is the time to start inserting details of the runtime compute node environment we use to run the tasks (I was directed not to do so at this stage, leaving it in the hands of TES as it is today).

Reply:
A thought just occurred to me. Would it make more sense to add a "prefix" property to the nodetask object, and let that be the environmentally-resolved prefix, rather than prefix all of the TES paths with a prefix containing an environment variable? If you agree, I'll do that first thing in the morning.